File: ruby/scrabble-score/scrabble_score.rb
# frozen_string_literal: true

# class Scrabble
class Scrabble
  def initialize(word)
    @word = word.nil? ? ' ' : word.upcase
  end

  def score
    points = 0
    ones = 'AEIOULNRST'
    twos = 'DG'
    threes = 'BCMP'
    fours = 'FHVWY'
    fives = 'K'
    eights = 'JX'
    tens = 'QZ'
    tiles = @word.split('')
    tiles.each do |tile|
      if ones.include?(tile)
        points += 1
      elsif twos.include?(tile)
        points += 2
      elsif threes.include?(tile)
        points += 3
      elsif fours.include?(tile)
        points += 4
      elsif fives.include?(tile)
        points += 5
      elsif eights.include?(tile)
        points += 8
      elsif tens.include?(tile)
        points += 10
      end
    end
    points
  end

  def self.score(word)
    new(word).score
  end
end

File: go/scrabble-score/scrabble_score.go
// Package scrabble does scrably things
package scrabble

import "strings"

var values = map[rune]int{}

// Score the letters submitted
func Score(word string) int {
	word = strings.ToLower(word)
	score := 0
	buildMap()
	for _, r := range word {
		score += values[r]
	}
	return score
}

func buildMap() {
	ones := []rune("aeioulnrst")
	twos := []rune("dg")
	threes := []rune("bcmp")
	fours := []rune("fhvwy")
	fives := []rune("k")
	eights := []rune("jx")
	tens := []rune("qz")

	for _, r := range ones {
		values[r] = 1
	}
	for _, r := range twos {
		values[r] = 2
	}
	for _, r := range threes {
		values[r] = 3
	}
	for _, r := range fours {
		values[r] = 4
	}
	for _, r := range fives {
		values[r] = 5
	}
	for _, r := range eights {
		values[r] = 8
	}
	for _, r := range tens {
		values[r] = 10
	}
}
